lms/events/event.py

```diff
@@ -101,3 +103,60 @@ def _get_assignment_id(self):
             return assignment.id
 
         return None
+
+
+@dataclass
+class AuditTrailEvent(BaseEvent):
+    type: EventType.Type = EventType.Type.AUDIT_TRAIL
+
+    instance: BASE = None
+    """Object for which we are tracking changes"""
+
+    action: str = None
+    """What happen to the object: crated,updated,deleted..."""
+
+    source: str = None
+    """In which context the change happen"""
+
+    def _get_data(self):
+        """
+        Fill the event's data value.
+
+        This is called on BaseEvent.__post_init__
+        """
+        changes = {}
+        instance_details = inspect(self.instance)
+        for attr in instance_details.attrs:
+            history = instance_details.get_history(attr.key, True)
+
+            if not history.has_changes():
+                continue
+
+            changes[attr.key] = (
+                history.deleted[0] if history.deleted else None,
+                history.added[0] if history.added else None,
+            )
+
+        return {
+            "model": self.instance.__class__.__name__,
+            "id": self.instance.id,
+            "action": self.action,
+            "source": self.source,
+            # LTI users will have a FK to the User table.
+            # userid is useful for other authentication methods.
+            # For example this will be the user's email while using google oauth.
+            "userid": self.request.identity.userid,
+            "changes": changes,
+        }
```

Comment:
I'm not suggesting this as a change here, as it's quite disruptive, but I think we might be fighting with two levels of nesting being flat here. It looks like we want a lot of attributes to be inside the data, but have attached them to the parent object.

If we have a separate object just for this we can make use of `asdict` on the object. That sub-object might look like this:

```python
@dataclass
class ModelChange:
    action: str
    model: str
    id: str
    source: str
    userid: str
    changes: Dict[str, tuple]

    @classmethod
    def from_instance(cls, instance, **kwargs):
        changes = {}
        instance_details = inspect(instance)
        for attr in instance_details.attrs:
            history = instance_details.get_history(attr.key, True)

            if not history.has_changes():
                continue

            changes[attr.key] = (
                history.deleted[0] if history.deleted else None,
                history.added[0] if history.added else None,
            )

        return cls(
            id=instance.id, model=instance.__class__.name, changes=changes, **kwargs
        )
```

This would save us the job of making the dict ourself and would clean up the parent object a lot:

```python
@dataclass
class AuditTrailEvent(BaseEvent):
    type: EventType.Type = EventType.Type.AUDIT_TRAIL

    change_data: ModelChange = None

    def _get_data(self):
        return asdict(self.change_data)
```

Reply:
I think this will also require the `"should_notify: bool"`

and then emit the event with `request.notify(AuditTrailEvent(request, ModelChange.from_instance(obj))`

I don't see any immediate benefits